doc/ipython-notebooks/multiclass/KNN.ipynb

```diff
@@ -260,14 +261,107 @@
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "Yall_low = np.genfromtxt('../../../data/multiclass/letter-recognition.dat', delimiter=',', usecols=(0,), dtype=\"string\")\n",
+    "Xall_low = np.genfromtxt('../../../data/multiclass/letter-recognition.dat', delimiter=',', usecols=(range(1,17))).transpose()\n",
+    "\n",
+    "# cover alphabet to numbers\n",
+    "Yall_low = np.array([ord(data) - 64 for data in Yall_low], dtype=np.double)\n",
+    "\n",
+    "Xtrain_low = Xall_low[:, :16000]\n",
+    "Ytrain_low = Yall_low[:16000]\n",
+    "\n",
+    "Xtest_low = Xall_low[:, 16000:]\n",
+    "Ytest_low = Yall_low[16000:]\n",
+    "\n",
+    "print Xtrain_low.shape\n",
+    "print Xtest_low.shape"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "labels_low = MulticlassLabels(Ytrain_low)\n",
+    "feats_low  = RealFeatures(Xtrain_low)\n",
+    "k=3\n",
+    "dist_low = EuclideanDistance()\n",
+    "knn_low = KNN(k, dist_low, labels_low)\n",
+    "labels_test_low = MulticlassLabels(Ytest_low)\n",
+    "feats_test_low  = RealFeatures(Xtest_low)\n",
+    "knn_low.train(feats_low)\n",
+    "pred_low = knn_low.apply_multiclass(feats_test_low)\n",
+    "print \"Predictions\", pred_low[:5]\n",
+    "print \"Ground Truth\", Ytest_low[:5]\n",
+    "\n",
+    "evaluator_low = MulticlassAccuracy()\n",
+    "accuracy_low = evaluator_low.evaluate(pred_low, labels_test_low)\n",
+    "\n",
+    "print \"Accuracy = %2.2f%%\" % (100*accuracy_low)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "start = Time.get_curtime()\n",
+    "knn_low.set_k(3)\n",
+    "knn_low.set_knn_solver_type(KNN_BRUTE)\n",
+    "pred_low = knn_low.apply_multiclass(feats_test_low)\n",
+    "print \"Standard KNN took %2.4fs\" % (Time.get_curtime() - start)\n",
+    "\n",
+    "\n",
+    "start = Time.get_curtime()\n",
+    "knn_low.set_k(3)\n",
+    "knn_low.set_knn_solver_type(KNN_COVER_TREE)\n",
+    "pred_low = knn_low.apply_multiclass(feats_test_low)\n",
+    "print \"Covertree KNN took %2.4fs\" % (Time.get_curtime() - start)\n",
+    "\n",
+    "\n",
+    "start = Time.get_curtime()\n",
+    "knn_low.set_k(3)\n",
+    "knn_low.set_knn_solver_type(KNN_KDTREE)\n",
+    "pred_low = knn_low.apply_multiclass(feats_test_low)\n",
+    "print \"KDTree KNN took %2.4fs\" % (Time.get_curtime() - start)"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "So we can significantly speed it up. Let's do a more systematic comparison. For that a helper function is defined to run the evaluation for KNN:"
+    "We can see that the cover tree solver is faster than standard KNN from above comparison. Next, let's do a more systematic comparison of Shogun's KNN implementations. We begin by defining a helper function:"
```

Comment:
Man, it doesn't make sense. Above you say that KD-trees are slower due to high dimensionality of the data, and now you run things on a low-dimensional dataset, KD-tree is still slower than cover tree.
You have to say something about this.

"We can see that the KD-tree solver now is faster than standard KNN on this 16-dimensional dataset. The cover tree solver is still fastest."

Or a bit more elaborate, hope you get the point